misc/lint/keep-sorted.sh

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# keep-sorted.sh — Checks that marked regions stay sorted.
+
+set -euo pipefail
+
+# shellcheck source=misc/shlib/shlib.bash
+. "$(dirname "$0")/../shlib/shlib.bash"
+
+if [[ $# -lt 1 ]]; then
+    echo "usage: $0 <file>" >&2
+    exit 1
+fi
+errors=0
+
+for file in "$@"; do
+    if [[ ! -f "$file" ]]; then
+        echo "lint: $(red error:) trailing-newline: internal error: $file is not a file" >&2
+        exit 1
+    fi
+
+    diff "$file" <(gawk '/\/\/ keep-sorted-start/ {f=1; delete a}
+      /\/\/ keep-sorted-end/ {f=0; n=asort(a); for (i=1;i<=n;i++) print a[i]}
+      !f
+      f{a[$0]=$0}' "$file") && error=$? || error=$?
+    if [ "$error" -gt 1 ]; then
+        echo "lint: $(red error:) diff internal error $error"
+        exit "$error"
+    elif [ "$error" -eq 1 ]; then
+        echo "lint: $(red error:) $file not sorted"
```

Comment:
```suggestion
        echo "lint: $(red error:) keep-sorted blocks in $file are not correctly sorted"
```

Reply:
It still uses `gawk` though! Can we change to `awk` now?

Reply:
Now it's not using gnu extensions anymore :) Took me a but to add a `--rewrite` option.